File: src/main/gnu/test/io/humble/video/RationalTest.cpp

// for isinf()
#include <math.h>

#include <io/humble/video/Global.h>
#include "RationalTest.h"

using namespace io::humble::ferry;

void
RationalTest :: setUp()
{
  num = 0;
}

void
RationalTest :: testCreationAndDestruction()
{
  num = Rational::make();
  VS_TUT_ENSURE("", num);
  VS_TUT_ENSURE_EQUALS("", num->getNumerator(), 0);
  VS_TUT_ENSURE_EQUALS("", num->getDenominator(), 1);

  num = Rational::make(6.0);
  VS_TUT_ENSURE("", num);
  VS_TUT_ENSURE_EQUALS("", num->getNumerator(), 6);
  VS_TUT_ENSURE_EQUALS("", num->getDenominator(), 1);

  num = Rational::make(6.1);
  VS_TUT_ENSURE("", num);
  VS_TUT_ENSURE_EQUALS("", num->getNumerator(), 61);
  VS_TUT_ENSURE_EQUALS("", num->getDenominator(), 10);
}

void
RationalTest :: testReduction()
{
  int retval = -1;
  num = Rational::make(2.2);
  VS_TUT_ENSURE("", num);
  VS_TUT_ENSURE_EQUALS("", num->getNumerator(), 11);
  VS_TUT_ENSURE_EQUALS("", num->getDenominator(), 5);

  retval = num->reduce(num->getNumerator()*5,
      num->getDenominator()*10, 100);
  VS_TUT_ENSURE("not exact", retval == 1);
  VS_TUT_ENSURE_EQUALS("", num->getNumerator(), 11);
  VS_TUT_ENSURE_EQUALS("", num->getDenominator(), 10);

  retval = Rational::sReduce(num.value(), 33, 32, 10);
  VS_TUT_ENSURE("exact?", retval == 0);
  VS_TUT_ENSURE_EQUALS("", num->getNumerator(), 1);
  VS_TUT_ENSURE_EQUALS("", num->getDenominator(), 1);

  // should be infinity
  retval = num->reduce(33, 0, 10);
  VS_TUT_ENSURE("not exact", retval == 1);
  VS_TUT_ENSURE_EQUALS("", num->getNumerator(), 1);
  VS_TUT_ENSURE_EQUALS("", num->getDenominator(), 0);
  VS_TUT_ENSURE("", isinf(num->getDouble()));

}

void
RationalTest :: testGetDouble()
{
  double retval = -1;
  num = Rational::make();

  retval = num->getDouble();
  VS_TUT_ENSURE_DISTANCE("", retval, 0, 0.0001);

  // now let make sure we can create infinity (and beyond...)
  num = Rational::make();
  num->reduce(1, 0, 10);
  retval = num->getDouble();
  VS_TUT_ENSURE("", isinf(retval));
}

void
RationalTest :: testMultiplication()
{
  RefPointer<Rational> a;
  RefPointer<Rational> b;

  a = Rational::make(12);
  b = Rational::make(3);
  num = Rational::sMultiply(a.value(), b.value());
  VS_TUT_ENSURE_DISTANCE("", num->getDouble(), 36, 0.0001);
}

void
RationalTest :: testAddition()
{
  RefPointer<Rational> a;
  RefPointer<Rational> b;

  a = Rational::make(12);
  b = Rational::make(3);
  num = Rational::sAdd(a.value(), b.value());
  VS_TUT_ENSURE_DISTANCE("", num->getDouble(), 15, 0.0001);
}

void
RationalTest :: testSubtraction()
{
  RefPointer<Rational> a;
  RefPointer<Rational> b;

  a = Rational::make(12);
  b = Rational::make(3);
  num = Rational::sSubtract(a.value(), b.value());
  VS_TUT_ENSURE_DISTANCE("", num->getDouble(), 9, 0.0001);
}

void
RationalTest :: testDivision()
{
  RefPointer<Rational> a;
  RefPointer<Rational> b;

  a = Rational::make(12);
  b = Rational::make(3);
  num = Rational::sDivide(a.value(), b.value());
  VS_TUT_ENSURE_DISTANCE("", num->getDouble(), 4, 0.0001);

  a = Rational::make(1);
  b = Rational::make(0.0);
  num = Rational::sDivide(a.value(), b.value());
  VS_TUT_ENSURE("", isinf(num->getDouble()));

  a = Rational::make(0.0);
  b = Rational::make(0.0);
  num = Rational::sDivide(a.value(), b.value());
  VS_TUT_ENSURE("", isnan(num->getDouble()));
}

void
RationalTest :: testConstructionFromNumeratorAndDenominatorPair()
{
  num = Rational::make(1, 10);
  VS_TUT_ENSURE_DISTANCE("", num->getDouble(), 0.1, 0.0001);
  num = Rational::make(2, 10);
  VS_TUT_ENSURE_DISTANCE("", num->getDouble(), 0.2, 0.0001);
  VS_TUT_ENSURE_EQUALS("", num->getNumerator(), 1);
  VS_TUT_ENSURE_EQUALS("", num->getDenominator(), 5);
}

void
RationalTest :: testRescaling()
{
  RefPointer<Rational> a;
  RefPointer<Rational> b;
  a = Rational::make(1, 100);
  VS_TUT_ENSURE_DISTANCE("", a->getDouble(), 0.01, 0.0001);
  b = Rational::make(1,5);
  VS_TUT_ENSURE_DISTANCE("", b->getDouble(), 0.2, 0.0001);

  VS_TUT_ENSURE_EQUALS("", a->rescale(1, b.value()), 20);
  VS_TUT_ENSURE_EQUALS("", b->rescale(1, a.value()), 0);
}

File: humble-video-all/src/test/java/io/humble/video/RationalTest.java
package io.humble.video;

import org.junit.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import io.humble.video.Rational;

import junit.framework.TestCase;

public class RationalTest extends TestCase
{
  private final Logger log = LoggerFactory.getLogger(this.getClass());

  private Rational mRational=null;

  @Before
  public void setUp()
  {
    log.debug("Executing test case: {}", this.getName());
    if (mRational != null)
      mRational.delete();
    mRational = null;
  }
  
  @Test
  public void testConstructorsAndGetters()
  {
    mRational = Rational.make();    
    assertTrue(mRational.getNumerator() == 0);
    assertTrue(mRational.getDenominator() == 1);
    mRational.delete();
    mRational = null;
    
    mRational = Rational.make(6.0);
    assertTrue(mRational.getNumerator() == 6);
    assertTrue(mRational.getDenominator() == 1);
    mRational.delete();
    mRational = null;

    mRational = Rational.make(6.1);
    assertTrue(mRational.getNumerator() == 61);
    assertTrue(mRational.getDenominator() == 10);
    mRational.delete();
    mRational = null;
  }
  
  @Test
  public void testReduction()
  {
    int retval = -1;
    mRational = Rational.make(2.2);
    
    assertTrue(mRational.getNumerator() == 11);
    assertTrue(mRational.getDenominator() == 5);

    retval = mRational.reduce(mRational.getNumerator()*5,
        mRational.getDenominator()*10, 100);
    assertTrue("not exact reduction", retval ==1);
    assertTrue(mRational.getNumerator()==11);
    assertTrue(mRational.getDenominator()==10);
    
    retval = Rational.sReduce(mRational, 33, 32, 10);
    assertTrue("exact reduction?", retval == 0);
    assertTrue(mRational.getNumerator()==1);
    assertTrue(mRational.getDenominator()==1);

    // A reduction to a den of 0 should be exact
    // but result in an 0 denominator and a 1
    // numerator (i.e. INFINITY).
    retval = mRational.reduce(33, 0, 10);
    assertTrue("exact reduction?", retval == 1);
    assertTrue(mRational.getNumerator()==1);
    assertTrue(mRational.getDenominator()==0);
  }
  
  @Test
  public void testCompareTo()
  {
    int retval = -1;
    Rational otherRational = Rational.make(5);
    mRational = Rational.make(4);

    retval = mRational.compareTo(otherRational);
    assertTrue(retval <0);
    retval = otherRational.compareTo(mRational);
    assertTrue(retval >0);
    retval = mRational.compareTo(mRational);
    assertTrue(retval == 0);
    retval = Rational.sCompareTo(mRational, otherRational);
    assertTrue(retval < 0);
  }
  
  @Test
  public void testGetDouble()
  {
    double retval = -1;
    mRational = Rational.make();
    retval = mRational.getDouble();
    assertTrue(retval == 0);
    
    // Now, let's create an invalid value and see what happens.
    mRational = Rational.make();
    mRational.reduce(1,0, 10);
    retval = mRational.getDouble();
    assertTrue("double is valid: " + retval, Double.isInfinite(retval));
  }
  
  @Test
  public void testMultiply()
  {
    Rational a = Rational.make(12);
    Rational b = Rational.make(3);
    
    mRational = Rational.sMultiply(a, b);
    assertTrue(mRational.getDouble() == 36);
  }

  @Test
  public void testAdd()
  {
    Rational a = Rational.make(12);
    Rational b = Rational.make(3);
    
    mRational = Rational.sAdd(a, b);
    assertTrue(mRational.getDouble() == 15);
  }  

  @Test
  public void testSubtract()
  {
    Rational a = Rational.make(12);
    Rational b = Rational.make(3);
    
    mRational = Rational.sSubtract(a, b);
    assertTrue(mRational.getDouble() == 9);
  }
  
  @Test
  public void testDivision()
  {
    Rational a = Rational.make(12);
    Rational b = Rational.make(3);
    
    mRational = Rational.sDivide(a, b);
    assertTrue(mRational.getDouble() == 4);
    
    mRational = Rational.sDivide(
        Rational.make(1),
        Rational.make(0));
    assertTrue(Double.isInfinite(mRational.getDouble()));

    mRational = Rational.sDivide(
        Rational.make(0),
        Rational.make(0));
    assertTrue(Double.isNaN(mRational.getDouble()));
  }
  
  @Test
  public void testCreateFromNumAndDen()
  {
    mRational = Rational.make(1, 10);
    assertTrue(mRational.getDouble() == 0.1);
    mRational = Rational.make(2,10);
    assertTrue(mRational.getDouble() == 0.2);
    assertTrue(mRational.getNumerator() == 1);
    assertTrue(mRational.getDenominator() == 5);
    
  }
}
